README.rst
=======
mnutree
=======

A json menu tree generator from csv file

The file generator comes with a mock data CSV that represents one of the many types of data that we have to deal with at retailshop.
The challenge is to consume and transform the CSV file in to a nested JSON file which will form a tree structure.

Description
===========
The following stack is used for the project
Python 3.8 (>3.8)
PyScaffold 3.2.3

The project can be installed to the python virtual environment or python host with command

  ``python -m pip install --upgrade pip``

  ``python setup.py install``

  This will install the commands to local system

  **for distrubation**

  ``pip install -U wheel``

  ``python setup.py dists``

  This will create a .whl distrubation under dist/ directory

  ``pip install dist/*.whl``

  This will install the commands to local system

  **for development purpose**

  ``pip install -r requirements.txt``

  ``pip install -e <path-to-mnutree>``

For example say **path-to-mnutree** is user's download folder at ``~/Downloads``.
The following can be executed.

with python
-----------
  ``$ cd  ~/Downloads/mnutree/``

  ``$ python setup.py install``

with pip
---------
  ``$ pip install -e ~/Downloads/mnutree``

for pytest
----------
  ``$ cd ~/Downloads/mnutree``

  **and fire command**

  ``$ python -m pytest -s --cov tests/``

  **or**

  ``$ py.test -s --cov tests/``

  **or**

  ``$ coverage run --source mnutree,mnuapi -m pytest``

  ``$ coverage report``

  **for BDD testing**

  ``$ python -m pytest -s -k "service" ``

  **for HTML report**

  ``$ pytest --html=report.html --self-contained-html``

for documentation
-----------------
  ``$ cd ~/Downloads/mnutree/``

  ``$ make -C docs html``

  the html file will be avilable under ``build/sphinx/html/``

  make command should be avilable in the system

for type checking
-----------------
  ``$ cd ~/Downloads/mnutree/``

  ``$ mypy src/``

  ``$ mypy --ignore-missing-imports tests/``


The ``$ python setup.py install`` will install the commands `mnutree` & `mnuapi`. Please type `mnutree -h` to see the options as:-
-----------------------------------------------------------------------------------------
usage: mnutree [-h] [--version] [-v] [-vv] [FILE]

The utility to convert structured csv to JSON

positional arguments:
  FILE                 input csv file with menu data

optional arguments:
  -h, --help           show this help message and exit
  --version            show program's version number and exit
  -v, --verbose        set loglevel to INFO
  -vv, --very-verbose  set loglevel to DEBUG

example
-------
  ``$ mnutree -v``

  The **mnutree** command can also takes the csv file as command line parameter like

  ``$ mnutree -v ~/Downloads/data.csv``

  **or**

  ``$ mnutree -v ~/Downloads/xyz.csv``

  ``$ mnuapi -v`` will start the api server. The swagger-ui is located at http://127.0.0.1:5000/docs
  Please press Ctrl+c to quit the server

Docker Local & Production
=========================
* The development server is started with ``$ docker-compose up``
* The production image is created locally with ``$ docker build . --target production -t mnutree``
* The production image is run as ``$ docker run -p 8080:8080 --env "PORT=8080" -it mnutree``

Note
====
The api uses uvicorn & gunicorn to serve high performance RESTFul requests.
A multi stage docker set up is created for the API.

This project has been set up using PyScaffold 3.2.3. For details and usage
information on PyScaffold see https://pyscaffold.org/.


CHANGELOG.rst
=========
Changelog
=========

Version 0.1
===========

- Changed project scaffolding
- Implemented the csv parsing


tests/__init__.py
# -*- coding: utf-8 -*-
"""The test module
"""
import sys

sys.path.append('..')


tests/conftest.py
# -*- coding: utf-8 -*-
"""The file used by pytest to auto-discover
   the fixtures and the  data shared amongst various tests.
   This file is included for demo purpose and it is not having a
   very strong use case in this module.
"""

from types import SimpleNamespace
import pytest

@pytest.fixture(scope="module")
def menu_item():
    """The method to show fixture returning
       just the dictionary
    """
    return {
        'label': 'CHEESE',
        'id': '178975',
        'link': 'https://groceries.morrisons.com/browse/178974/178969/178975'
    }


@pytest.fixture(scope="module")
def cmd_arg():
    """The method to return the command line
       arguments for testing purpose
    """
    cmd_input_path = "tests/data.csv"
    args = {"csv_file": cmd_input_path, "loglevel":"20"}

    return SimpleNamespace(**args)


tests/test_main.py
# -*- coding: utf-8 -*-
"""The test script for the main module
"""
from pathlib import Path
from mnutree.main import main

def test_main(cmd_arg):
    """The method to test main
    """
    main(cmd_arg)

    if cmd_arg.csv_file.find("/") > -1:
        file_path = Path(cmd_arg.csv_file)
    else:
        file_path = Path("./"+cmd_arg.csv_file)

    file_path = file_path.with_suffix('.json')

    assert file_path.exists()
    assert file_path.is_file()


src/mnutree/main.py
# -*- coding: utf-8 -*-
"""This is a main file that can serve as a starting point for a Python console script.
   console_scripts =
        mnutree = mnutree.main:run

   Then run `python setup.py install` which will install the
   command `mnutree` inside your current environment.
   Besides console scripts, the header (i.e. until logger...) of this module can
   also be used as template for Python modules.
"""
import sys
import json
from pathlib import Path
from types import SimpleNamespace
from typing import List, TextIO, Dict, Union, Any

from mnutree import info
from mnutree import parse_args
from mnutree import setup_logging
from mnutree.processor import process

__author__ = "Gaurav J"
__copyright__ = "Mastek India Pvt. Ltd"
__license__ = "MIT"

def main(args: Any) -> None:
    """Main entry point allowing external calls

       Parameters
       ----------
       args: ([str])
        command line parameter list
    """
    if not isinstance(args, SimpleNamespace):
        args = parse_args(args)

    setup_logging(args.loglevel)
    info("Starting JSON conversion...")

    menus: List[Dict[str, Union[str, Any]]]
    file_path: Path
    json_file: TextIO

    file_path , menus  = process(args)

    menus_str: str = json.dumps(menus)

    with file_path.open("w") as json_file:
        json_file.writelines(menus_str)

    info("Conversion to JSON ends here")

def run() -> None:
    """Entry point for console_scripts
    """

    main(sys.argv[1:])

if __name__ == "__main__":
    run()


src/mnutree/skeleton.py
# -*- coding: utf-8 -*-
"""
This is a skeleton file that can serve as a starting point for a Python
console script. To run this script uncomment the following lines in the
[options.entry_points] section in setup.cfg:

    console_scripts =
         fibonacci = mnutree.skeleton:run

Then run `python setup.py install` which will install the command `fibonacci`
inside your current environment.
Besides console scripts, the header (i.e. until _logger...) of this file can
also be used as template for Python modules.

Note: This skeleton file can be safely removed if not needed!
"""

import argparse
import sys
import logging

from mnutree import __version__

__author__ = "Gaurav J"
__copyright__ = "Gaurav J"
__license__ = "mit"

_logger = logging.getLogger(__name__)


def fib(n):
    """Fibonacci example function

    Args:
      n (int): integer

    Returns:
      int: n-th Fibonacci number
    """
    assert n > 0
    a, b = 1, 1
    for i in range(n-1):
        a, b = b, a+b
    return a


def parse_args(args):
    """Parse command line parameters

    Args:
      args ([str]): command line parameters as list of strings

    Returns:
      :obj:`argparse.Namespace`: command line parameters namespace
    """
    parser = argparse.ArgumentParser(
        description="Just a Fibonacci demonstration")
    parser.add_argument(
        "--version",
        action="version",
        version="mnutree {ver}".format(ver=__version__))
    parser.add_argument(
        dest="n",
        help="n-th Fibonacci number",
        type=int,
        metavar="INT")
    parser.add_argument(
        "-v",
        "--verbose",
        dest="loglevel",
        help="set loglevel to INFO",
        action="store_const",
        const=logging.INFO)
    parser.add_argument(
        "-vv",
        "--very-verbose",
        dest="loglevel",
        help="set loglevel to DEBUG",
        action="store_const",
        const=logging.DEBUG)
    return parser.parse_args(args)


def setup_logging(loglevel):
    """Setup basic logging

    Args:
      loglevel (int): minimum loglevel for emitting messages
    """
    logformat = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"
    logging.basicConfig(level=loglevel, stream=sys.stdout,
                        format=logformat, datefmt="%Y-%m-%d %H:%M:%S")


def main(args):
    """Main entry point allowing external calls

    Args:
      args ([str]): command line parameter list
    """
    args = parse_args(args)
    setup_logging(args.loglevel)
    _logger.debug("Starting crazy calculations...")
    print("The {}-th Fibonacci number is {}".format(args.n, fib(args.n)))
    _logger.info("Script ends here")


def run():
    """Entry point for console_scripts
    """
    main(sys.argv[1:])


if __name__ == "__main__":
    run()


src/mnutree/processor.py
# -*- coding: utf-8 -*-
"""A module to read the csv file containig the menu data.
   The module than form the parent child relation ship in csv data.
   The module gives a list containing dictoniaries. The dictonry
   follow the parent child structure as derived from the csv menu file.
"""
import re
from pathlib import Path
from argparse import Namespace
from collections import defaultdict
from typing import List, Tuple, Dict, DefaultDict, Optional, Any

from mnutree import info, strfile, FILE_STR

__author__ = "Gaurav J"
__copyright__ = "Mastek India Pvt. Ltd"
__license__ = "MIT"

KEYS = "label", "id", "link"

def normallize(line: str) -> str:
    """The function currently removes extra quotes and replaces comma.
       The function uses regex to find a comma in quotes.
       The extra comma in csv file is an issue as it prevents
       the separation of the coloums.

       Parameters
       ----------
       line : string
        the line as read from the csv file
    """
    quotes: List[str] = re.findall(r'"(.*?)"', line)
    if len(quotes) != 0:
        for quote in quotes:
            line = line.replace(quote, quote.replace(",", "\u02cc"))
            line = line.replace("\"", "")

    info("normallize: {}", line, debug=True)
    return line


def process(args: Namespace) -> Tuple[Path, List[Dict[str, Any]]]:
    """The main method doing heavy lifting.
       The method to process the csv file
       to the JSON tree structure as required

       Parameters
       ----------
       args : dictionary
        the csv file path from command line

       Returns
       -------
       tuple
        the file path & parent-child dictionary of menus
    """
    if args.csv_file.find("/") > -1:
        file_path: Path = Path(args.csv_file)
    else:
        file_name: str
        with strfile(FILE_STR) as file_name:
            file_path = Path(file_name)

    menu_list: List[Dict[str, Any]] = []
    object_cache: DefaultDict[str, Dict[str, Any]] = defaultdict(lambda : {})

    with file_path.open() as file:
        line: str
        line_count: int = 0
        for line in file:
            line = line.rstrip()
            if line_count == 0:
                line_count += 1
                info("process:header: {} ", line, debug=True)
            else:
                line_count += 1
                line = normallize(line)
                list_of_items_in_line: List[str] = [word for word in line.split(',')
                    if word and word != "\n"]
                len_list_of_items_in_line: int = len(list_of_items_in_line)
                info("process: {} . len --> {}, list_of_items_in_line --> {}",
                    line_count, len_list_of_items_in_line, list_of_items_in_line, debug=True)
                if len_list_of_items_in_line != 0:
                    create_menu(list_of_items_in_line, menu_list, object_cache)


    if file_path.name.find("tmp") > -1:
        info("process: The file path is {} ", file_path)
        info("data: {} ", menu_list)
    else:
        info("process: The file path is {} ", file_path.with_suffix('.json'))

    return file_path.with_suffix('.json'), menu_list


def create_menu(list_of_items_in_line: List[str], menu_list: List[Dict[str, Any]],
     object_cache: DefaultDict[str, Dict[str, Any]]) -> None:
    """The method implementing the algo.
       The method takes the list of list_of_items_in_line (menu items)
       and transforms them into lists & dictionary. It
       then insert the childeren at appropriate places.

       Parameters
       ----------
       list_of_items_in_line : list
        the list of menu item as obtained from single line of csv file

       menu_list : list
        the main root list which will finally have json replica

       object_cache : dictionary
        the dictionary to hold the menu items
    """
    info("create_menu: {}", menu_list, debug=True)

    list_of_items_in_line.pop(0)
    item_to_group: int = 3
    item_group_list: List[List[str]] = [list_of_items_in_line[i:i + item_to_group]
        for i in range(0, len(list_of_items_in_line), item_to_group)]

    word_list: List[str]
    for word_list in item_group_list:
        menu_item: Dict[str, Any] = {key: word.replace("\u02cc", ",") if key == KEYS[0] else word
            for key, word in zip(KEYS, word_list)}

        pid: Optional[str] = parent_id(menu_item, KEYS[2], KEYS[1])

        if pid is None:
            cached_item: Dict[str, Any] = object_cache[menu_item[KEYS[1]]]
            cached_item.setdefault(KEYS[0], menu_item[KEYS[0]])
            cached_item.setdefault(KEYS[1], menu_item[KEYS[1]])
            cached_item.setdefault(KEYS[2], menu_item[KEYS[2]])

            root_item: Optional[Dict[str, Any]] = find(cached_item[KEYS[1]], menu_list)
            info("root item --> {}", root_item, debug=True)
            if not root_item:
                info("create_menu:root: {} --> {}", menu_item[KEYS[1]], menu_item)
                menu_list.append(cached_item)
        else:
            info("create_menu:child: {} --> {}", menu_item[KEYS[1]], menu_item, debug=True)

            cached_item = object_cache[menu_item[KEYS[1]]]
            cached_item.setdefault(KEYS[0], menu_item[KEYS[0]])
            cached_item.setdefault(KEYS[1], menu_item[KEYS[1]])
            cached_item.setdefault(KEYS[2], menu_item[KEYS[2]])

            parent: Dict[str, Any] = object_cache[pid]
            parent.setdefault(KEYS[1], pid)
            parent.setdefault("children", [cached_item])

            child: Optional[Dict[str, Any]] = find(cached_item[KEYS[1]], parent["children"])
            if not child:
                parent["children"].append(cached_item)

def parent_id(menu_item: Dict[str, Any], split_attribute: str,
    match_attribute: str) -> Optional[str]:
    """The method to extract parent id.
       The method extracts the parent id from the link
       The last id is the menu item id and the secon last id
       is the parent id. This parent id is returned or None

       Parameters
       ----------
       menu_item : dictionary
        the representation of the menu item

       split_attribute : string
        the attribute which is to be split for parent id extraction

       match_attribute : string
        the attribute for cross verification of the item

       Returns
       -------
       int
        the id of the parent menu item
    """
    link = menu_item[split_attribute].split("/")
    pid: str = link[-2]
    if pid.isdigit() and menu_item[match_attribute] == link[-1]:
        info("parent_id: found parent id {}", pid, debug=True)
        return pid

    return None

def find(item_id: str, items: List[Dict[str, Any]]) -> Optional[Dict[str, Any]]:
    """The method to search item in the list.
       The method finds an item in the list given it's id

       Parameters
       ----------
       item_id : string
        the id of the menu item
       items : list
        the list to check for

       Returns
       -------
       dictionary
        the representation of the menu item or None
    """
    if len(items) == 0:
        return None

    for item in items:
        if item[KEYS[1]] == item_id:
            return item
    return None

src/mnuapi/req_processor.py
# -*- coding: utf-8 -*-
"""The router for menu tree request processing
"""
from io import BytesIO
from collections import defaultdict
from typing import List, Dict, DefaultDict, Any

from fastapi import status, Response
from fastapi.routing import APIRouter
from fastapi.param_functions import File
from fastapi.datastructures import UploadFile
from fastapi.exceptions import HTTPException

from mnutree import info
from mnutree.processor import normallize, create_menu

router: APIRouter = APIRouter()

@router.post("/process", tags=["process-csv"], responses={
    200: {"description": "file is processed"},
    400: {"description": "bad request"},
    500: {"description": "internal server error"}
})
async def accept_file(response: Response, uploaded_file: UploadFile = File(...)) -> Dict[str, Any]:
    """The method to take file multi-part upload.
       The file is uploaded to the memory
    """
    menu_list: List[Dict[str, Any]] = []

    try:
        if uploaded_file.content_type != "text/csv":
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                detail="Please provide csv file")

        contents: Any = await uploaded_file.read()

        if len(contents) == 0:
            raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,
                detail="The uploaded csv file is empty")

        file: BytesIO = BytesIO(contents)

        object_cache: DefaultDict[str, Dict[str, Any]] = defaultdict(lambda : {})

        line_count: int = 0
        for line in file:
            line = line.rstrip()
            if line_count == 0:
                line_count += 1
                info("process:header: {} ", line, debug=True)
            else:
                line_count += 1
                line_str: str = normallize(line.decode('utf-8'))
                list_of_items_in_line: List[str] = [word for word in line_str.split(',')
                    if word and word != "\n"]
                len_list_of_items_in_line: int = len(list_of_items_in_line)
                info("process: {} . len --> {}, list_of_items_in_line --> {}",
                    line_count, len_list_of_items_in_line, list_of_items_in_line, debug=True)
                if len_list_of_items_in_line != 0:
                    create_menu(list_of_items_in_line, menu_list, object_cache)

    except IOError as error:
        response.status_code = status.HTTP_500_INTERNAL_SERVER_ERROR
        return {
            "message": f'an error occured {error}'
        }

    return {
        "message":
            f'The file {uploaded_file.filename} of '
            f'type {uploaded_file.content_type} processed successfully',
        "data": menu_list
    }


src/mnutree/__init__.py
# -*- coding: utf-8 -*-
"""The menu tree module file.
   The file is auto-generated by pyscaffolding
   The file is modified to include logging construct which
   is common across all files.
"""
import os
import sys
import logging
from logging import Logger, getLogger
from argparse import ArgumentParser, Namespace
from typing import List, Tuple, Sequence, Optional, TextIO, Generator
import tempfile
from contextlib import contextmanager

from pkg_resources import get_distribution, DistributionNotFound

try:
    # Change here if project is renamed and does not equal the package name
    DIST_NAME: str = __name__
    __version__: str = get_distribution(DIST_NAME).version
except DistributionNotFound:
    __version__ = 'unknown'
finally:
    del get_distribution, DistributionNotFound

def is_valid_file(parser: ArgumentParser, arg: str) -> Optional[str]:
    """The method to check if the file path is valid file.
       The file path is provided from command line
    """
    if not os.path.exists(arg) and arg.find("/") > -1:
        parser.error("The menu file %s does not exist!" % arg)
        return None
    else:
        if arg.find("/") > -1:
            data_file: TextIO
            with open(arg, 'r') as data_file:
                if not data_file.readable:
                    parser.error("The menu file %s is not readable!" % arg)
                    return None
    return arg


def parse_args(args: Optional[Sequence[str]]) -> Namespace:
    """Parse command line parameters

       Parameters
       ----------
       args: ([str])
        command line parameters as list of strings

       Returns
       -------
       :obj:`argparse.Namespace`
        command line parameters namespace
    """
    parser: ArgumentParser = ArgumentParser(description
        ="The utility to convert structured csv to JSON")

    parser.add_argument(dest="csv_file", help="input csv file with menu data",
                        metavar="FILE", default="data.csv", nargs='?',
                        const=1, type=lambda x: is_valid_file(parser, x))

    parser.add_argument("--version", action="version",
                        version="menucsv {ver}".format(ver=__version__),
    )

    parser.add_argument("-v", "--verbose",
                        dest="loglevel", help="set loglevel to INFO",
                        action="store_const", const=logging.INFO,
    )

    parser.add_argument("-vv", "--very-verbose",
                        dest="loglevel", help="set loglevel to DEBUG",
                        action="store_const", const=logging.DEBUG,
    )

    return parser.parse_args(args)


def setup_logging(loglevel: int) -> None:
    """Setup basic logging

       Parameters
       ----------
       loglevel: int
        minimum loglevel for emitting messages
    """
    logformat: str = "[%(asctime)s] %(levelname)s:%(name)s:%(message)s"

    logging.basicConfig(
        level=loglevel, stream=sys.stdout, format=logformat, datefmt="%Y-%m-%d %H:%M:%S"
    )

__logger__ : Logger = getLogger(__name__)

# the method is to be evolved furthur
def info(message: str, *values: object, debug: bool=False) -> None:
    """The unified logging construct

      The method to control the logging in the application.

      Parameters
      ----------
      message : string
          the logging message

      *values: non-keyworded, variable-length argument list
            the values used in the format method the string

      debug: boolean
            the default is false, so log level is info by default
            true will enable debug level
    """
    if len(values) == 0:
        if debug:
            __logger__.debug(message)
        else: __logger__.info(message)
    else:
        if debug :
            __logger__.debug(message.format(*values))
        else:
            __logger__.info(message.format(*values))


@contextmanager
def strfile(data: str) -> Generator:
    """The temp file context manager
    """
    temp = tempfile.NamedTemporaryFile(delete=False)
    config_path = f'{temp.name}.csv'
    with open(config_path, 'w') as file:
        file.write(data)
    try:
        yield f'{temp.name}.csv'
    finally:
        temp.close()
        os.unlink(temp.name)


FILE_STR = """Base URL,Level 1 - Name,Level 1 - ID,Level 1 - URL,Level 2 - Name,Level 2 - ID,\
    Level 2 URL,Level 3 - Name,Level 3 - ID,Level 3 URL
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,,,,,,
,,,,,,,,,
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,,,
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,CHEESE,178975,https://groceries.morrisons.com/browse/178974/178969/178975
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,COOKED MEAT & ANTIPASTI,178976,https://groceries.morrisons.com/browse/178974/178969/178976
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,DAIRY & EGGS,178977,https://groceries.morrisons.com/browse/178974/178969/178977
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,DESSERTS,178978,https://groceries.morrisons.com/browse/178974/178969/178978
https://groceries.morrisons.com/browse,THE BEST,178974,https://groceries.morrisons.com/browse/178974,FRESH,178969,https://groceries.morrisons.com/browse/178974/178969,DRINKS,178979,https://groceries.morrisons.com/browse/178974/178969/178979
"""